components/validator/issuer.go

```diff
@@ -0,0 +1,272 @@
+	// The candidacy block might be issued and fail to be accepted in the node for various reasons,
+	// so we should stop issuing candidacy blocks only when the account is successfully registered as a candidate.
+	// For this reason,
+	// retry interval parameter should be bigger than the fishing threshold to avoid issuing redundant candidacy blocks.
+	executor.ExecuteAt(CandidateTask, func() { candidateAction(ctx) }, now.Add(ParamsValidator.CandidacyRetryInterval))
+
+	if !deps.NodeBridge.NodeStatus().GetIsBootstrapped() {
+		Component.LogDebug("not issuing candidate block because node is not bootstrapped yet.")
+
+		return
+	}
+
+	if err = issueCandidateBlock(ctx, now, currentAPI); err != nil {
+		Component.LogWarnf("error while trying to issue candidacy announcement: %s", err.Error())
+	}
+
+}
+
+func committeeMemberAction(ctx context.Context) {
+	now := time.Now()
+	currentAPI := deps.NodeBridge.APIProvider().APIForTime(now)
+	currentSlot := currentAPI.TimeProvider().SlotFromTime(now)
+	currentEpoch := currentAPI.TimeProvider().EpochFromSlot(currentSlot)
+
+	isCommitteeMember, err := deps.NodeBridge.ReadIsCommitteeMember(ctx, validatorAccount.ID(), currentSlot)
+	if err != nil {
+		Component.LogWarnf("error while checking if account %s is a committee member in slot %d: %s", validatorAccount.ID(), currentSlot, err.Error())
+		executor.ExecuteAt(CommitteeTask, func() { committeeMemberAction(ctx) }, now.Add(ParamsValidator.CommitteeBroadcastInterval))
+
+		return
+	}
+
+	if !isCommitteeMember {
+		Component.LogDebug("account %s is not a committee member in epoch %d", currentEpoch)
+		executor.ExecuteAt(CommitteeTask, func() { committeeMemberAction(ctx) }, currentAPI.TimeProvider().SlotStartTime(currentAPI.TimeProvider().EpochStart(currentEpoch+1)))
+
+		return
+	}
+
+	// Schedule next committeeMemberAction regardless of whether the node is bootstrapped or validator block is issued
+	// as it must be issued as part of validator's responsibility.
+	executor.ExecuteAt(CommitteeTask, func() { committeeMemberAction(ctx) }, now.Add(ParamsValidator.CommitteeBroadcastInterval))
+
+	// Validator block may ignore the bootstrap flag in order to bootstrap the network and begin acceptance.
+	if !ParamsValidator.IgnoreBootstrapped && !deps.NodeBridge.NodeStatus().GetIsBootstrapped() {
+		Component.LogDebug("not issuing validator block because node is not bootstrapped yet.")
+
+		return
+	}
+
+	if err = issueValidatorBlock(ctx, now, currentAPI); err != nil {
+		Component.LogWarnf("error while trying to issue candidacy announcement: %s", err.Error())
+	}
+}
+
+func issueCandidateBlock(ctx context.Context, blockIssuingTime time.Time, currentAPI iotago.API) error {
+	blockSlot := currentAPI.TimeProvider().SlotFromTime(blockIssuingTime)
+
+	strongParents, weakParents, shallowLikeParents, err := deps.NodeBridge.RequestTips(ctx, iotago.BlockMaxParents)
+	if err != nil {
+		return ierrors.Wrapf(err, "failed to get tips for candidate block")
```

Comment:
Maybe remove the "candidate block" in all the error messages here, because it's added outside of the function anyway.